File: apps/mobile/ios/Visvine/Repositories/ProfileRepository.swift
import Foundation

struct ProfileRepository {
    private let api = APIClient.shared

    func getProfile(personId: String) async -> APIResult<DirectoryMember> {
        let res: APIResult<DirectoryMember> = await api.request("/api/profile/\(encoded(personId))")
        return resolveMember(res)
    }

    func getFullProfile(personId: String) async -> APIResult<FullProfile> {
        let res: APIResult<FullProfile> = await api.request("/api/profile/\(encoded(personId))")
        switch res {
        case .success(var profile):
            profile.imageUrl = MediaURL.resolve(profile.imageUrl)
            return .success(profile)
        case .failure(let m):
            return .failure(m)
        }
    }

    func updateProfile(userId: String, update: ProfileUpdate) async -> APIResult<DirectoryMember> {
        let body = try? JSONEncoder().encode(update)
        let res: APIResult<DirectoryMember> = await api.request(
            "/api/profile/\(encoded(userId))", method: "PATCH", body: body
        )
        return resolveMember(res)
    }

    private func resolveMember(_ res: APIResult<DirectoryMember>) -> APIResult<DirectoryMember> {
        switch res {
        case .success(var member):
            member.imageUrl = MediaURL.resolve(member.imageUrl)
            return .success(member)
        case .failure(let m):
            return .failure(m)
        }
    }

    private func encoded(_ id: String) -> String {
        id.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? id
    }
}

File: apps/mobile/android/app/src/main/java/com/visvine/mobile/data/repository/ProfileRepository.kt
package com.visvine.mobile.data.repository

import com.visvine.mobile.data.model.DirectoryMember
import com.visvine.mobile.data.model.FullProfile
import com.visvine.mobile.data.model.ProfileUpdate
import com.visvine.mobile.data.remote.ApiResult
import com.visvine.mobile.data.remote.MediaUrl
import com.visvine.mobile.data.remote.VisvineApi
import com.visvine.mobile.data.remote.safeApiCall
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ProfileRepository @Inject constructor(
    private val api: VisvineApi,
    private val json: Json,
) {
    suspend fun getProfile(personId: String): ApiResult<DirectoryMember> =
        when (val res = safeApiCall(json) { api.getProfile(personId) }) {
            is ApiResult.Success -> ApiResult.Success(res.data.copy(imageUrl = MediaUrl.resolve(res.data.imageUrl)))
            is ApiResult.Failure -> res
        }

    suspend fun getFullProfile(personId: String): ApiResult<FullProfile> =
        when (val res = safeApiCall(json) { api.getFullProfile(personId) }) {
            is ApiResult.Success -> ApiResult.Success(res.data.copy(imageUrl = MediaUrl.resolve(res.data.imageUrl)))
            is ApiResult.Failure -> res
        }

    suspend fun updateProfile(userId: String, update: ProfileUpdate): ApiResult<DirectoryMember> =
        when (val res = safeApiCall(json) { api.updateProfile(userId, update) }) {
            is ApiResult.Success -> ApiResult.Success(res.data.copy(imageUrl = MediaUrl.resolve(res.data.imageUrl)))
            is ApiResult.Failure -> res
        }
}
